plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/ExpEmbedding.kt

```diff
@@ -129,71 +144,76 @@ sealed interface BinaryBooleanExpression : ExpEmbedding {
 }
 
 data class Not(
     val exp: ExpEmbedding,
+    override val source: KtSourceElement? = null,
 ) : ExpEmbedding {
     override val type = BooleanTypeEmbedding
 
-    override fun toViper() = Exp.Not(exp.toViper())
+    override fun toViper() = Exp.Not(exp.toViper(), source.asSilverPosition)
 }
 
 
 data object UnitLit : ExpEmbedding {
     override val type = UnitTypeEmbedding
+    override val source: KtSourceElement? = null
 
     override fun toViper() = UnitDomain.element
 }
 
-data class IntLit(val value: Int) : ExpEmbedding {
+data class IntLit(val value: Int, override val source: KtSourceElement? = null) : ExpEmbedding {
     override val type = IntTypeEmbedding
 
     override fun toViper() = Exp.IntLit(value)
 }
 
-data class BooleanLit(val value: Boolean) : ExpEmbedding {
+data class BooleanLit(val value: Boolean, override val source: KtSourceElement? = null) : ExpEmbedding {
     override val type = BooleanTypeEmbedding
 
     override fun toViper() = Exp.BoolLit(value)
 }
 
-data class NullLit(val elemType: TypeEmbedding) : ExpEmbedding {
+data class NullLit(val elemType: TypeEmbedding, override val source: KtSourceElement? = null) : ExpEmbedding {
     override val type = NullableTypeEmbedding(elemType)
-
-    override fun toViper() = NullableDomain.nullVal(elemType.viperType)
+    override fun toViper() = NullableDomain.nullVal(elemType.viperType, source)
 }
 
-data class Is(val exp: ExpEmbedding, val comparisonType: TypeEmbedding) : ExpEmbedding {
+data class Is(val exp: ExpEmbedding, val comparisonType: TypeEmbedding, override val source: KtSourceElement? = null) : ExpEmbedding {
     override val type = BooleanTypeEmbedding
 
     override fun toViper() =
         TypeDomain.isSubtype(TypeOfDomain.typeOf(exp.toViper()), comparisonType.runtimeType)
 }
 
-data class Cast(val exp: ExpEmbedding, override val type: TypeEmbedding) : ExpEmbedding {
-    override fun toViper() = CastingDomain.cast(exp.toViper(), type)
+data class Cast(val exp: ExpEmbedding, override val type: TypeEmbedding, override val source: KtSourceElement? = null) : ExpEmbedding {
+    override fun toViper() = CastingDomain.cast(exp.toViper(), type, source)
     override fun ignoringCasts(): ExpEmbedding = exp
 }
 
-data class FieldAccess(val receiver: ExpEmbedding, val field: FieldEmbedding) : ExpEmbedding {
+data class FieldAccess(val receiver: ExpEmbedding, val field: FieldEmbedding, override val source: KtSourceElement? = null) : ExpEmbedding {
     override val type: TypeEmbedding = field.type
-    override fun toViper() = Exp.FieldAccess(receiver.toViper(), field.toViper())
-    fun getAccessPredicate(perm: PermExp = PermExp.FullPerm()) = AccessPredicate.FieldAccessPredicate(toViper(), perm)
+    override fun toViper() = Exp.FieldAccess(receiver.toViper(), field.toViper(), source.asSilverPosition)
+    fun getAccessPredicate(perm: PermExp = PermExp.FullPerm()) =
+        AccessPredicate.FieldAccessPredicate(toViper(), perm, source.asSilverPosition)
 }
```

Comment:
I'm a bit confused about how we get a position for a field access permission.  These are typically entirely synthetic: e.g. we produce a field access permission for a field of an object passed as a parameter.  What source position does it get?

Reply:
I see what you mean, then the `getAccessPredicate` function should take a source parameter to me. I would suggest renaming the function from `get...` to `createAccessPredicate`.

WDYT?

Reply:
Probably a solution would be the following one:
```kotlin
    override fun toViper() = Exp.FieldAccess(receiver.toViper(), field.toViper(), source.asPosition)
    fun getAccessPredicate(perm: PermExp, source: KtSourceElement?) =
        AccessPredicate.FieldAccessPredicate(toViper().copy(pos = source.asPosition), perm, source.asPosition)
```

Reply:
I don't have a strong opinion on the name.

In complicated cases like this it might be helpful to provide a comment with an explanation of what ends up referring to which position, especially with something like an access predicate which is not actually present in the original code.